versioned_hdf5/subchunk_map.py

```diff
@@ -36,294 +45,353 @@
-    start = (start - s_start) // s_step
+class DropAxis(enum.Enum):
+    _drop_axis = 0
 
-    stop: Py_ssize_t = min(c_stop, s_stop)
-    stop = ceil_a_over_b(stop - s_start, s_step) if stop > s_start else 0
 
-    return slice(int(start), int(stop), 1)
+# Returned instead of an AnySlicer. Signals that the axis should be removed when
+# aggregated into an AnySlicerND.
+DROP_AXIS = DropAxis._drop_axis
 
 
-@cython.cfunc
-def _subindex_slice_chunk(
-    s_start: Py_ssize_t,
-    s_stop: Py_ssize_t,
-    s_step: Py_ssize_t,
-    c_start: Py_ssize_t,
-    c_stop: Py_ssize_t,
-) -> slice:
-    """Given a slice(s_start, s_stop, s_step) indexing an axis of
-    a dataset, return a slice that's been clipped to select only
-    the data within the range [c_start, c_stop[ and shifted back
-    by s_start.
-
-    See examples on _subindex_chunk_slice
+@cython.cclass
+class IndexChunkMapper:
+    """Abstract class that manipulates a numpy fancy index along a single axis of a
+    chunked array
+
+    Parameters
+    ----------
+    chunk_indices:
+        Array of indices of all the chunks involved in the selection along the axis
+    chunk_size:
+        Size of each chunk, in points, along the axis
+    dset_size:
+        Size of the whole array, in points, along the axis
     """
-    start: Py_ssize_t = max(s_start, c_start)
-    # Get the smallest step multiple of common that is >= start
-    start = smallest_step_after(start, s_start % s_step, s_step)
-    # Finally, we need to shift start so that it is relative to index
-    start -= c_start
-
-    stop: Py_ssize_t = max(0, min(s_stop, c_stop) - c_start)
-
-    # This is the same, in the special case we're in, to
-    #     return Slice(start, stop, s_step).reduce(d).raw
-    # It's reimplemented here for speed.
-    # assert 0 <= start < stop <= d
-    step: Py_ssize_t = s_step
-    if start + step >= stop:
-        stop, step = start + 1, 1  # Indexes 1 element
-    else:
-        stop -= (stop - start - 1) % step
-        if stop - start == 1:
-            step = 1  # Indexes 1 element
-    return slice(int(start), int(stop), int(step))
 
+    chunk_indices: hsize_t[:]
+    dset_size: hsize_t
+    chunk_size: hsize_t
+
+    def __init__(
+        self,
+        chunk_indices: hsize_t[:],
+        dset_size: hsize_t,
+        chunk_size: hsize_t,
+    ):
+        self.chunk_indices = chunk_indices
+        self.dset_size = dset_size
+        self.chunk_size = chunk_size
+
+    @cython.cfunc
+    @cython.nogil
+    @cython.exceptval(check=False)
+    def _chunk_start_stop(self, chunk_idx: hsize_t) -> tuple[hsize_t, hsize_t]:
+        """Return the range of points [a, b[ of the chunk indexed by chunk_idx"""
+        if not cython.compiled:
+            chunk_idx = int(chunk_idx)  # np.uint64 + int -> np.float64
+
+        start = chunk_idx * self.chunk_size
+        stop = min(start + self.chunk_size, self.dset_size)
+        return start, stop
+
+    @cython.ccall
+    @abc.abstractmethod
```

Comment:
Is it a problem that `IndexChunkMapper` doesn't currently inherit from `abc.ABCMeta` when using [this decorator](https://docs.python.org/3/library/abc.html#abc.abstractmethod)?

Reply:
It is, actually. It means that if you forget to override an abstractmethod nothing bad happens and you end up with a dummy method that returns None.
However, cython breaks if you try either `class IndexChunkMapper(abc.ABC)` or `class IndexChunkMapper(metaclass=abc.ABCMeta)`. So the `abc.abstractmethod` decorator remains purely descriptive. 

*sigh*.
I'll write a note about it.